### DISCUSSION TITLE
TypeError: cloudPlugin is not a function

### DISCUSSION THREAD

**Discussion by Abcmsaj** [currents-dev/cypress-currents-reporter#65] [Q&A] [answered]
Forgive me if I'm being dumb here - looking to rip out currents/cli and migrate to cypress-cloud. I have a config file in the root directory with record key and project ID in it, pointing to Currents URL. Inside our config file for our environment, I have added 

`import { defineConfig } from 'cypress';
const cloudPlugin = require('cypress-cloud/plugin');`

at the top and my e2e section looks like this (because we have other tasks in here):

`e2e: {
    setupNodeEvents(on, config) {
      on('before:browser:launch', (browser, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.preferences.default.intl = { accept_languages: 'en-GB' };
          launchOptions.args.push('--auto-open-devtools-for-tabs');
          return launchOptions;
        }
      });

      on('task', {
        /* Task to read an Excel file */
        readExcelFile(filename) {
          return readXlsxFile(filename);
        },
        /* Task to get the last download from the downloads folder (order by time) */
        getLastDownload(downloadsPath) {
          // If downloads folder doesn't exist, return null
          if (!existsSync(downloadsPath)) {
            return null;
          }

          // Order files by datetime
          const filesOrdered = readdirSync(downloadsPath)
            .map((entry: any) => path.join(downloadsPath, entry))
            .filter((entryWithPath: any) => lstatSync(entryWithPath).isFile())
            .map((fileName: any) => ({ fileName, mtime: lstatSync(fileName).mtime }))
            .sort((a: any, b: any) => b.mtime.getTime() - a.mtime.getTime());

          // If there are no files, return null
          if (!filesOrdered.length) {
            return null;
          }

          // Get the latest file in the list
          return filesOrdered[0].fileName;
        }
      });

      return cloudPlugin(on, config);
    },
    baseUrl: '<url here>,
    supportFile: 'cypress/support/e2e.js',
    specPattern: 'cypress/e2e/**/**/*.{js,jsx,ts,tsx}',

  },`

but I get an error when it runs in CLI

`'',
    '[STARTED] Task without title.',
    '[SUCCESS] Task without title.',
    '',
    'Opening Cypress...',
    'Your configFile threw an error from: /azp/_work/1/s/AutomationRepoEndToEndTests/cypress/config/int.config.ts',
    '',
    'The error was thrown while executing your e2e.setupNodeEvents() function:',
    '',
    'TypeError: cloudPlugin is not a function',
    '    at setupNodeEvents (/azp/_work/1/s/AutomationRepoEndToEndTests/cypress/config/int.config.ts:90:14)',
    '    at /root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:118:14',
    '    at tryCatcher (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/bluebird/js/release/util.js:16:23)',
    '    at Function.Promise.attempt.Promise.try (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/bluebird/js/release/method.js:39:29)',
    '    at RunPlugins.load (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:115:9)',
    '    at RunPlugins.runSetupNodeEvents (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:255:10)',
    '    at EventEmitter.<anonymous> (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:185:22)',
    '    at EventEmitter.emit (node:events:513:28)',
    '    at EventEmitter.emit (node:domain:489:12)',
    '    at process.<anonymous> (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:33:22)',
    '    at process.emit (node:events:513:28)',
    '    at process.emit (node:domain:489:12)',
    '    at process.emit.sharedData.processEmitHook.installedValue [as emit] (/root/.cache/Cypress/12.7.0/Cypress/resources/app/node_modules/@cspotcode/source-map-support/source-map-support.js:745:40)',
    '    at emit (node:internal/child_process:937:14)',
    '    at processTicksAndRejections (node:internal/process/task_queues:83:21)',
    ''
  ],
  stderr: [
    '[466:0309/112729.760500:ERROR:zygote_host_impl_linux.cc(263)] Failed to adjust OOM score of renderer with pid 619: Permission denied (13)',
    '[619:0309/112729.778588:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.',
    'Missing baseUrl in compilerOptions. tsconfig-paths will be skipped',
    ''
  ]
}

====================================================================================================

    ERROR  Error: Unable to resolve cypress configuration
    at bootCypress (/azp/_work/1/s/AutomationRepoEndToEndTests/node_modules/cypress-cloud/lib/bootstrap.ts:81:11)
    at getConfig (/azp/_work/1/s/AutomationRepoEndToEndTests/node_modules/cypress-cloud/lib/config.ts:58:19)
    at run (/azp/_work/1/s/AutomationRepoEndToEndTests/node_modules/cypress-cloud/index.ts:45:18)`

Any help appreciated if I'm doing something wrong - thanks

**Comment by Abcmsaj** [comment]
Another Q I had was why is there no --headless argument when running cypress-cloud? is it headless by default?

**Reply by agoldis** [reply]
It runs in headless mode by default

**Comment by arendjantetteroo** [comment]
If you are in a typescript environment for your cypress config file, use this instead:
`import { cloudPlugin } from 'cypress-cloud/plugin';`

**Comment by Abcmsaj** [comment]
![Screenshot 2023-03-09 at 11 56 11](https://user-images.githubusercontent.com/5143023/224015960-b768bad5-32d7-4135-97b3-424b7eb5f527.png)

I am using a ts file for config so this would make sense, however, with this line, I now get this - have checked and the module is installed on my machine, present in node_modules dir

**Reply by Abcmsaj** [reply]
Ah I changed the config file to js and that seems to work now - except:

   Discovered 36 spec files
   Tags: atportal; Group: false; Parallel: true; Batch Size: 3
   Connecting to cloud orchestration service...
    WARNING  Network request failed: 'connect ECONNREFUSED X.X.X.X:80'. Next attempt is in 15s (1/3).
    WARNING  Network request failed: 'connect ECONNREFUSED X.X.X.X:80'. Next attempt is in 30s (2/3).
    WARNING  Network request failed: 'connect ECONNREFUSED X.X.X.X:80'. Next attempt is in 1m (3/3).

(dont know what the IP is so I omitted it!)

**Reply by arendjantetteroo** [reply]
I had the same about type declarations and put the following above that line. 
It just means typescript can't find the definitions for the plugin. 
```
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```
I added this issue for it: https://github.com/currents-dev/cypress-cloud/issues/66

For the connect, it seems you need to check your director URL to see if that's correct and reachable from your system.

**Reply by Abcmsaj** [reply]
Connect was just coz I copied and pasted an example config that had http instead of https! easy fix - I'll change all our env configs back to ts if this works and add that comment as you suggested - hopefully that'll all be good!! Thanks for your help

**Reply by agoldis** [reply]
Also, the TS exports were fixed in 1.4.2

**Comment by agoldis** [comment, marked as answer]
Hey, sorry, we messed up the released package. Please upgrade to 1.4.2 and use 

```
const {cloudPlugin} = require("cypress-cloud/plugin");
```
